source/extensions/common/wasm/context.cc

```diff
@@ -0,0 +1,216 @@
+
+#include "common/common/assert.h"
+#include "common/common/empty_string.h"
+#include "common/common/enum_to_int.h"
+#include "common/common/logger.h"
+
+#include "extensions/common/wasm/wasm.h"
+#include "extensions/common/wasm/wasm_state.h"
+#include "extensions/common/wasm/well_known_names.h"
+
+#include "absl/base/casts.h"
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/node_hash_map.h"
+#include "absl/strings/str_cat.h"
+#include "absl/synchronization/mutex.h"
+#include "openssl/bytestring.h"
+#include "openssl/hmac.h"
+#include "openssl/sha.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+
+bool Buffer::copyTo(WasmBase* wasm, size_t start, size_t length, uint64_t ptr_ptr,
+                    uint64_t size_ptr) const {
+  if (start + length > data_.size()) {
+    return false;
+  }
+  absl::string_view s = data_.substr(start, length);
+  return wasm->copyToPointerSize(s, ptr_ptr, size_ptr);
+}
+
+Wasm* Context::wasm() const { return static_cast<Wasm*>(wasm_); }
+
+void Context::error(absl::string_view message) { throw WasmException(std::string(message)); }
+
+void Context::initializeRoot(WasmBase* wasm, PluginBaseSharedPtr plugin) {
+  ContextBase::initializeRoot(wasm, plugin);
+  root_local_info_ = &std::static_pointer_cast<Plugin>(plugin)->local_info_;
+}
+
+uint64_t Context::getCurrentTimeNanoseconds() {
+  return std::chrono::duration_cast<std::chrono::nanoseconds>(
+             wasm()->time_source_.systemTime().time_since_epoch())
+      .count();
+}
+
+WasmResult Context::log(uint64_t level, absl::string_view message) {
+  switch (static_cast<proxy_wasm::LogLevel>(level)) {
+  case proxy_wasm::LogLevel::trace:
+    ENVOY_LOG(trace, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::debug:
+    ENVOY_LOG(debug, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::info:
+    ENVOY_LOG(info, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::warn:
+    ENVOY_LOG(warn, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::error:
+    ENVOY_LOG(error, "wasm log{}: {}", log_prefix(), message);
+    break;
+  case proxy_wasm::LogLevel::critical:
+    ENVOY_LOG(critical, "wasm log{}: {}", log_prefix(), message);
+    break;
+  default:
+    return WasmResult::BadArgument;
+  }
+  return WasmResult::Ok;
+}
+
+WasmResult Context::getProperty(absl::string_view path, std::string* result) {
+  if (path.size() == 15 && !memcmp(reinterpret_cast<const void*>(&*path.begin()),
+                                   reinterpret_cast<const void*>("plugin_root_id\0"), 15)) {
+    *result = std::string(root_id());
+    return WasmResult::Ok;
+  }
+  return WasmResult::Unimplemented;
+}
+
+WasmResult Context::defineMetric(MetricType type, absl::string_view name, uint32_t* metric_id_ptr) {
+  auto stat_name = wasm()->stat_name_set_->add(name);
```

Comment:
Note that this call to StatNameSet::add() will consume memory unconditionally, even if the you call it with the same name on the same set multiple times.

StatNameSet::rememberBuiltin() is the way to get re-use via a set. Just be aware that that method is not thread-safe.

Reply:
Also note that it will unconditionally take a global symbol-table lock on every call, which I think is also not what you want, unless this is the first time that name has ever been seen.

Reply:
OK, I'll put a lock on.  At this point it looks like memory is more of an issue than efficiency.

Reply:
A lock and a change to call rememberBuiltin() rather than add()?

Reply:
So, what is the advantage of the complex tokenizing system if we still have a StatNameSet which has a map from full strings to the tokenized representation?  Doesn't that use more memory?

Reply:
It uses significantly less memory in most scenarios, and it will also use less storage here if there are multiple calls to defineMetric() with the same string. We'd only allocate on the first call.

I think we can also make it more effective in the context of WASM but maybe not local to this code. Ping me offline; I think I've written you privately how I think this should be structured and I'm happy to dig that up again.

But for a localized change here I'd recommend using StatNameDynamicStorage locally here in lieu of StatNameSet:
```
  Stats::StatNameDynamicStorage stat_name_storage(name, wasm->scope_->symbolTable());
  Stats::StatName stat_name = stat_name_storage.statName();
```

Reply:
This discussion has moved offline and will be resolved in https://github.com/envoyproxy/envoy-wasm/pull/452 from which it will be upstreamed to here.